### PR TITLE
Julia/discord supported channels

### LINKED ISSUE #58536: Support notifications for different channel types in Discord
- Discord Forums need higher permissions and will say that message is sent when it does not send
- When permissions fail, maybe link to docs of all the necessary permissions for different use cases
```[tasklist]
### Tasks
- [ ] Update docs
```

### PR DESCRIPTION
**Pull request by 17hogeju**
Adding support for threads and giving specific error messages with failure.

Screenshot 2023-10-25 at 4 14 44 PM

Screenshot 2023-10-25 at 4 14 33 PM

Screenshot 2023-10-24 at 9 32 00 AM

Avoids case where user can send a notification to a forum directly and the notification tells user it sent, but it does not.
https://discord.com/channels/621778831602221064/804769304255266918/1164913320470515782

Closes https://github.com/getsentry/sentry/issues/58536

### CODE REVIEW

**Comment by 17hogeju** [comment]
Adding permission updates would mean that new integrations will have the proper permissions and existing ones will not. Old integrations must enable these permissions in settings. We will probably want a troubleshooting doc that directs users to check the permissions for their channel.

**Comment by 17hogeju** [comment]
Adding another case to the validation method that checks the type here should be ok... The case where it wouldn't be ok is if the user saved a rule where the channel is a forum.

**Comment by andrzejkaszkowiak** [comment]
Hi Julia, just a quick question.

Do you aim to support sending notifications to threads with this PR?

I tried pointing my alerts to a thread using `<channel_id>?thread_id=<thread_id>`* in the `channel id` field and sending test alerts, and it seemed to work on Sentry side, but nothing got posted to the thread. (Messages to channels works.)

*GitHub webhooks use this form

**Comment by 17hogeju** [comment]
@andrzejkaszkowiak yes, that is the intention. It will also let the user know that an error occurred if an unsupported type is used (forum) before creating the rule or sending a test notification.

**Comment by andrzejkaszkowiak** [comment]
Great, @17hogeju, thank you! Do you happen to have an ETA on this feature?
I'm not trying to pressure you — I want to estimate some things on my end.

File: src/sentry/integrations/discord/client.py [hunk 1]
```diff
@@ -18,6 +18,12 @@ class DiscordClient(IntegrationProxyClient):
     integration_name: str = "discord"
     base_url: str = "https://discord.com/api/v10"
 
+    # https://discord.com/developers/docs/resources/channel#channel-object-channel-types
+    GUILD_TEXT = 0
```
Comment (scefali) [inline src/sentry/integrations/discord/client.py:22]: @17hogeju can we do this with an enum?

File: src/sentry/integrations/discord/integration.py [hunk 2]
```diff
@@ -105,8 +105,13 @@ class DiscordIntegrationProvider(IntegrationProvider):
...
     # https://discord.com/developers/docs/topics/permissions#permissions
-    # Permissions value that can Send Messages (0x800), View Channel (0x400), and Embed Links (0x4000):
-    bot_permissions = 0x800 | 0x400 | 0x4000
+    # View Channel + read messages in text channels (1 << 10)
+    # Send Messages + create threads in forum (1 << 11)
+    # Embed Links (1 << 14)
+    # Create public threads (1 << 35)
+    # Create private threads (1 << 36)
+    # Send messages in threads (1 << 38)
+    bot_permissions = 1 << 10 | 1 << 11 | 1 << 14 | 1 << 35 | 1 << 36 | 1 << 38
```
Comment (scefali) [inline src/sentry/integrations/discord/integration.py:114]: @17hogeju Can we use enums for the bit shifts?

**Review by scefali — APPROVED** [review]

**Comment by 17hogeju** [comment]
@andrzejkaszkowiak Threads are now supported! If you don't want to reinstall your Sentry integration to update the bot permissions, you can manually add them within Discord. Go to Server Settings > Roles > Sentry > Permissions and enable `Send Messages in Threads`, `Create Public Threads` and `Create Private Threads`.

When you create an alert in Sentry, follow the same process for other channels as outlined in the [docs](https://docs.sentry.io/product/integrations/notification-incidents/discord/#issue-alerts).

If your thread is private, you will have to @ mention the sentry bot to add it to the thread.

**Comment by andrzejkaszkowiak** [comment]
I appreciate you notifying me. Thanks much! I will re-enable my integration to use the feature. Kudos, Julia!

**Pull request merged**